File: modules/foundation/src/main/java/com/projecthub/base/cohort/infrastructure/persistence/CohortPersistenceAdapter.java
package com.projecthub.base.cohort.infrastructure.persistence;

import com.projecthub.base.cohort.application.port.out.LoadCohortPort;
import com.projecthub.base.cohort.application.port.out.SaveCohortPort;
import com.projecthub.base.cohort.domain.entity.Cohort;
import com.projecthub.base.school.cohort.application.port.out.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.UUID;

@Component
@RequiredArgsConstructor
class CohortPersistenceAdapter implements LoadCohortPort, SaveCohortPort {
    private final CohortRepository cohortRepository;

    @Override
    public Optional<Cohort> loadCohortById(UUID id) {
        return cohortRepository.findById(id);
    }

    @Override
    public Page<Cohort> loadAllCohorts(Pageable pageable) {
        return cohortRepository.findAll(pageable);
    }

    @Override
    public Page<Cohort> loadCohortsBySchool(UUID schoolId, Pageable pageable) {
        return cohortRepository.findBySchoolId(schoolId, pageable);
    }

    @Override
    public Cohort save(Cohort cohort) {
        return cohortRepository.save(cohort);
    }

    @Override
    public void delete(Cohort cohort) {
        cohortRepository.deleteById(UUID);
    }
}

File: src/main/kotlin/com/projecthub/infrastructure/persistence/cohort/CohortPersistenceAdapter.kt
package com.projecthub.infrastructure.persistence.cohort

import com.projecthub.application.cohort.mapper.CohortMapper
import com.projecthub.domain.cohort.Cohort
import com.projecthub.domain.cohort.CohortRepository
import com.projecthub.infrastructure.persistence.cohort.jpa.CohortJpaRepository
import org.springframework.stereotype.Component

/**
 * Adapter implementing the CohortRepository interface
 * This adapter connects the domain layer to the persistence infrastructure
 */
@Component
class CohortPersistenceAdapter(
    private val cohortJpaRepository: CohortJpaRepository,
    private val cohortMapper: CohortMapper
) : CohortRepository {

    override fun findById(id: String): Cohort? {
        return cohortJpaRepository.findById(id)
            .map { cohortMapper.toDomainEntity(it) }
            .orElse(null)
    }

    override fun findAll(): List<Cohort> {
        return cohortJpaRepository.findAll().map { cohortMapper.toDomainEntity(it) }
    }

    override fun findBySchool(schoolId: String): List<Cohort> {
        return cohortJpaRepository.findBySchoolId(schoolId).map { cohortMapper.toDomainEntity(it) }
    }

    override fun save(cohort: Cohort): Cohort {
        val jpaEntity = cohortMapper.toJpaEntity(cohort)
        val savedEntity = cohortJpaRepository.save(jpaEntity)
        return cohortMapper.toDomainEntity(savedEntity)
    }

    override fun deleteById(id: String) {
        cohortJpaRepository.deleteById(id)
    }

    override fun findActiveCohortsWithCapacity(minCapacity: Int): List<Cohort> {
        return cohortJpaRepository.findActiveCohortsWithCapacity(minCapacity = minCapacity)
            .map { cohortMapper.toDomainEntity(it) }
    }
}
